nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/CaptureChangeMySQL.java

```diff
@@ -0,0 +1,1001 @@
+            }
+        }
+
+        doStop.set(false);
+    }
+
+
+    public void outputEvents(ProcessSession session, StateManager stateManager, ComponentLog log) throws IOException {
+        RawBinlogEvent rawBinlogEvent;
+
+        // Drain the queue
+        while ((rawBinlogEvent = queue.poll()) != null && !doStop.get()) {
+            Event event = rawBinlogEvent.getEvent();
+            EventHeaderV4 header = event.getHeader();
+            long timestamp = header.getTimestamp();
+            EventType eventType = header.getEventType();
+            // Advance the current binlog position. This way if no more events are received and the processor is stopped, it will resume at the event about to be processed.
+            // We always get ROTATE and FORMAT_DESCRIPTION messages no matter where we start (even from the end), and they won't have the correct "next position" value, so only
+            // advance the position if it is not that type of event. ROTATE events don't generate output CDC events and have the current binlog position in a special field, which
+            // is filled in during the ROTATE case
+            if (eventType != EventType.ROTATE && eventType != EventType.FORMAT_DESCRIPTION) {
+                currentBinlogPosition = header.getPosition();
+            }
+            log.debug("Got message event type: {} ", new Object[]{header.getEventType().toString()});
+            switch (eventType) {
+                case TABLE_MAP:
+                    // This is sent to inform which table is about to be changed by subsequent events
+                    TableMapEventData data = event.getData();
+
+                    // Should we skip this table? Yes if we've specified a DB or table name pattern and they don't match
+                    skipTable = (databaseNamePattern != null && !databaseNamePattern.matcher(data.getDatabase()).matches())
+                            || (tableNamePattern != null && !tableNamePattern.matcher(data.getTable()).matches());
+
+                    if (!skipTable) {
+                        TableInfoCacheKey key = new TableInfoCacheKey(this.getIdentifier(), data.getDatabase(), data.getTable(), data.getTableId());
+                        if (cacheClient != null) {
+                            try {
+                                currentTable = cacheClient.get(key, cacheKeySerializer, cacheValueDeserializer);
+                            } catch (ConnectException ce) {
+                                throw new IOException("Could not connect to Distributed Map Cache server to get table information", ce);
+                            }
+
+                            if (currentTable == null) {
+                                // We don't have an entry for this table yet, so fetch the info from the database and populate the cache
+                                try {
+                                    currentTable = loadTableInfo(key);
```

Comment:
Yeah this is an unfortunate limitation of not keeping a versioned history of the table schemas. Not sure what to do if we don't parse the DDL, and even then we'd need a full history. For example if we didn't enable CDC (binary logging) until after some tables were created, then made changes and dropped the table, then neither solution would give us what the table looked like at that point in time.